### PR TITLE
Add LLVM pipeline tests

### PR DESCRIPTION
**Pull request by pchintalapudi**
We test our passes in LLVM in isolation, but it doesn't look like we test our pass pipeline anywhere except nanosoldier's benchmarks. The tests here aren't benchmarks, but simply basic syntactical features we expect should be optimized in some way by our current pass pipeline. They should ideally prevent major regressions in the pipeline (e.g. LLVM version upgrade, adding/removing/shifting passes, switch to NewPM).

### CODE REVIEW

**Comment by pchintalapudi** [comment]
In this case I'm going for regressing on the syntax, rather than on the IR, so that we see our passes are having an effect on real code.

**Comment by vtjnash** [comment]
Can you add an explicit `Base.Experimental.@compiler_options optimize=2` at the top here?

We usually run tests with `--check-bounds=yes`, so I am not sure how this will work there. You might need an `_exec.jl` file, like we do for some of the other unfortunate tests.

**Comment by pchintalapudi** [comment]
I split everything out into O3, O2, O1, O0 test files, but since I don't have a clear definition of O3 and O1 beyond >= O2/O0, respectively, I just left those as testing only the levels below them.

File: test/llvmpasses/pipeline-o2.jl [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    for i=one(typeof(N)):N
+        total += i
+    end
+    total
+end
+# ALL-LABEL: @julia_simd_sumloop
+# ALL: vector.body
+function simd_sumloop(N)
+    total = zero(typeof(N))
+    @simd for i=one(typeof(N)):N
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o2.jl:114]:
```suggestion
    @simd for i in one(typeof(N)):N
```

File: test/llvmpasses/pipeline-o2.jl [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+end
+
+# COM: check hoisting and loop deletion functionality
+
+# ALL-LABEL: @julia_loopedlength
+# ALL-NOT: br
+# ALL: ret
+function loopedlength(arr)
+    len = length(arr)
+    for i=1:length(arr)
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o2.jl:127]:
```suggestion
    for i in 1:length(arr)
```

File: test/llvmpasses/pipeline-o2.jl [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+# BC_OFF: llvm.memset
+# BC_AUTO: llvm.memset
+
+# COM: check reductive indvars/vectorization
+
+# ALL-LABEL: @julia_sumloop
+# ALL: mul
+function sumloop(N)
+    total = zero(typeof(N))
+    for i=one(typeof(N)):N
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o2.jl:105]:
```suggestion
    for i in one(typeof(N)):N
```

File: test/llvmpasses/pipeline-o2-allocs.jl [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        ref = Ref(i)
+    end
+    ref
+end
+
+# CHECK-LABEL: @julia_loop_const
+# CHECK-NOT: br
+function loop_const()
+    ref = Ref(0)
+    for i=1:1000
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o2-allocs.jl:34]:
```suggestion
    for i in 1:1000
```

File: test/llvmpasses/pipeline-o2-allocs.jl [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    end
+end
+
+# CHECK-LABEL: @julia_loop_alloc
+# CHECK: phi
+# CHECK-NOT: phi
+function loop_alloc(N)
+    ref = Ref(zero(typeof(N)))
+    N <= zero(typeof(N)) && return ref
+    for i=one(typeof(N)):N
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o2-allocs.jl:24]:
```suggestion
    for i in one(typeof(N)):N
```

File: test/llvmpasses/pipeline-o0.jl [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+function simple()
+    Ref(0)
+end
+
+# CHECK-LABEL: @julia_buildarray
+# CHECK-NOT: julia.write_barrier
+# CHECK: gc_queue_root
+function buildarray()
+    out = []
+    for i=1:100
```
Comment (vchuravy) [inline test/llvmpasses/pipeline-o0.jl:23]:
```suggestion
    for i in 1:100
```

**Review by vchuravy — APPROVED** [review]

**Pull request merged**